calico_kubernetes/calico_kubernetes.py

```diff
@@ -181,32 +227,43 @@ def _container_add(self, pid, ip, interface, hostname, orchestrator_id):
         # Obtain information from Docker Client and validate container state
         self._validate_container_state(self.docker_id)
 
-        # Assign ip address through IPAM Client
-        try:
-            ip_assigned = self._datastore_client.assign_address(None, ip)
-        except PoolNotFound:
-            logger.error("IP address %s does not belong to any configured "
-                         "pools. Exiting.", ip)
-            sys.exit(1)
+        if CALICO_IPAM == 'calico':
+            logger.info("Using Calico IPAM")
+            # Assign ip address through IPAM Client
+            try:
+                ip_lists = self._datastore_client.auto_assign_ips(
+                1, 0, str(self.docker_id), None)
+            except RuntimeError as err:
+                logger.error("Cannot auto assign IPAddress: %s", err.message)
+
+            # Concatenate lists into one single list
+            ip_list = ip_lists[0]+ip_lists[1]
         else:
-            if not ip_assigned:
-               logger.error("Failed to assign IP address %s. Exiting.", ip)
-               sys.exit(1)
+            logger.info("Using docker assigned IP address")
+            ip_list = [self._read_docker_ip()]
+            try:
+                self._datastore_client.assign_ip(ip_list[0], self.docker_id, None)
+            except RuntimeError as err:
+                logger.error("Cannot assign IPAddress %s: %s", ip_list[0], err.message)
+                sys.exit(1)
 
         # Create Endpoint object
         try:
+            logger.info("Creating endpoint for ips in list %s", ip_list)
             ep = self._datastore_client.create_endpoint(hostname, orchestrator_id,
-                                                        self.docker_id, [ip])
+                                                        self.docker_id, ip_list)
         except AddrFormatError:
             logger.error("This node is not configured for IPv%d. Unassigning "
-                         "IP address %s then exiting.", ip.version, ip)
-            self._datastore_client.unassign_address(None, ip)
+                         "IP address %s then exiting.", ip_list[0].version, ip_list[0])
+            self._datastore_client.release_ips(set(ip_list))
             sys.exit(1)
 
         # Create the veth, move into the container namespace, add the IP and
         # set up the default routes.
-        ns = netns.PidNamespace(pid)
-        ep.mac = ep.provision_veth(ns, interface)
+        logger.info("Creating the veth with pid %s on interface %s", pid, interface)
```

Comment:
nitpick - the veth doesn't have a PID, it is the namespace's PID.

So, this log should be "Creating veth in namespace pid=%s with interface name %s" % (pid, interface)